rotkehlchen/chain/ethereum/uniswap/uniswap.py

```diff
@@ -106,6 +119,90 @@ def __init__(
                 f'Probably will get fixed with time. If not report it to Rotkis support channel ',
             )
 
+    @staticmethod
+    def _calculate_events_balances(
+            address: ChecksumEthAddress,
+            events: List[LiquidityPoolEvent],
+            balances: List[LiquidityPool],
+    ) -> List[LiquidityPoolEventsBalance]:
+        """Given an address and its LP events and LPs, process each LP event
+        (grouped by pool) aggregating the token0, token1, usd and LP token
+        amounts. Factorise in the aggregation the current protocol balances
+        (if `balances` != [], all events case). Finally return the profit/loss
+        totals and the LP events (grouped by pool) in
+        <LiquidityPoolEventsBalance>.
+        """
+        events_balances: List[LiquidityPoolEventsBalance] = []
+        pool_balance: Dict[ChecksumEthAddress, LiquidityPool] = (
+            {pool.address: pool for pool in balances}
+        )
+        # quick lookup, `agg` from aggregated
+        pool_events_agg_balance: Dict[ChecksumEthAddress, Dict[str, Any]] = {}
```

Comment:
Don't use dicts for that. It makes it hard to understand and use typing for. Use a dataclass or a named tuple. Since here it's mutable you will need a dataclass.

Also in the following code why do you use simple FVal for some of the stuff that seem to be a combination of amount + usd value? You know we have a structure for that: https://github.com/rotki/rotki/blob/c0e01b71a8da1de2debb50e91f992add599013fe/rotkehlchen/accounting/structures.py#L104

Perhaps it does not apply to this code. I am not 100% clear yet due to the late hour

Oh and please change the name. I do mean it when I say that I hate abbreviations. If you have to add a comment to tell me this means aggregation, then it's already telling you it's not a good name.

What's a better name? Well you don't need to namespace stuff.

`pool_events_agg_balance` -> `aggregated_balance`

Reply:
Using `dataclass`: sure, I'll put the values of the aggregated balances dict in a dataclass.

Using `Balance`: as far as I'm aware I can't use it because the events return the involved `token0` and `token1` amounts and the total USD price (resulting from SUM(token0 amount in USD, token1 amount in USD)). For using Balance per token I'd need to query each tokens prices per each event timestamp.